app/src/main/kotlin/io/homeassistant/companion/android/widgets/BaseGlanceEntityWidgetReceiver.kt

```diff
@@ -0,0 +1,284 @@
+import io.homeassistant.companion.android.common.data.servers.ServerManager
+import io.homeassistant.companion.android.database.widget.WidgetDao
+import javax.inject.Inject
+import kotlinx.coroutines.CoroutineExceptionHandler
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.Job
+import kotlinx.coroutines.SupervisorJob
+import kotlinx.coroutines.cancel
+import kotlinx.coroutines.isActive
+import kotlinx.coroutines.launch
+import timber.log.Timber
+
+private fun newCoroutineScopeProvider(): () -> CoroutineScope {
+    val exceptionHandler = CoroutineExceptionHandler { _, throwable -> Timber.e(throwable, "Unhandled error in widget scope") }
+    // Use SupervisorJob to avoid cancelling all jobs when one fails since the scope is used across all the widgets
+    return { CoroutineScope(Dispatchers.Main + SupervisorJob() + exceptionHandler) }
+}
+
+data class EntitiesPerServer(val serverId: Int, val entityIds: List<String>)
+
+/**
+ * Base class for Glance widgets that update based on entity state changes.
+ *
+ * This class provides the foundational functionality for managing widget updates, handling lifecycle events,
+ * and observing entity state changes. It is designed to be extended by specific widget implementations.
+ *
+ * ### Key Features:
+ * - **Entity state observation**: Watches for updates to specific entities and triggers widget updates accordingly.
+ * - **Lifecycle management**: Handles widget lifecycle action [Intent.ACTION_SCREEN_ON], [Intent.ACTION_SCREEN_OFF] and [AppWidgetManager.ACTION_APPWIDGET_UPDATE].
+ * - **Real-time updates**: Supports real-time updates for widgets when the user grants the necessary permissions.
+ * - **Cleanup**: Cancel subscription when widget are removed and make sure the database reflect the current widgets used.
+ *
+ * ### Notes:
+ * - Glance widget, may "freeze" subscriptions after some time unless the user grants
+ *   the `real-time` permission. To unfreeze, call `update` on the widget.
+ * - Ensure that data sources are observed within the composition state (e.g., using [androidx.compose.runtime.collectAsState]) to keep
+ *   widgets updated while the composition is active (less than a minute even with the `real-time` permission).
+ * - Make sur to invoke [registerReceiver] to register actions to manage widget updates effectively.
+ * - Even if we are watching for entity changes in the receiver the data received cannot be sent to the widget directly,
+ *   if you need to send data to the widget you need to update the [DAO] and call [update].
+ * - The changes are watch only after [Intent.ACTION_SCREEN_ON]. It means that on startup the widget won't watch for changes
+ *   until the screen is turn on and off.
+ *
+ * ### Usage:
+ * Extend this class and implement the abstract method `getWidgetEntitiesByServer` to provide the mapping of widget IDs
+ * to their associated entities.
+ *
+ * ### Example:
```

Comment:
Should this heading be level 4, or maybe removed as the next example doesn't have a heading?